eng/emitter-package.json

```diff
@@ -0,0 +1,11 @@
+{
+    "main": "dist/src/index.js",
+    "dependencies": {
+        "@azure-tools/typespec-rust": "0.1.0"
+    },
+    "devDependencies": {
+        "@azure-tools/typespec-client-generator-core": ">=0.46.2 <1.0.0",
+        "@typespec/compiler": ">=0.60.1 <1.0.0",
+        "@typespec/xml": ">=0.60.1 <1.0.0"
```

Comment:
NOTE: I manually added this for storage, however we can remove if it's not needed for now. This should be added in the peerDependencies of the typespec-rust package for it to get carried over to emitter-package.json whenever you use the script to sync dependencies.

Reply:
Do all tsp packages need to be listed as peer dependencies in an emitter's `package.json`? I thought it was only the ones the emitter directly uses, but I could be wrong.

Reply:
That is the convention the emitters have been following, this allows emitters to pin supported peer dependency ranges. We used to just manually pin these deps in emitter-package.json but I believe @hallipr worked on shifting to using the peer deps in the emitter's package.json so he could share more. From a tsp-client perspective we simply install whatever deps are defined in emitter-package.json. Examples:
[typespec-python](https://github.com/Azure/autorest.python/blob/f3f6491ca1d653b584b092b07f7907237cc00c4e/packages/typespec-python/package.json#L47-L58)
[typespec-csharp](https://github.com/Azure/autorest.csharp/blob/b11bb94bc90e29694ab222e3b30cb43f167cb949/src/TypeSpec.Extension/Emitter.Csharp/package.json#L41-L53)

Reply:
These shouldn't be specified as range dependencies.  We started pinning any floating peer dependencies with dev dependencies to reduce the risk of supply chain vulnerabilities caused by picking up a version in CI that we haven't yet verified on a dev machine.  Even though the emitter-package-lock.json will pin us to a specific version, we can't control which version we pin to without specifying the version in devDependencies.

`npm install` will automatically satisfy deep peer dependencies. This means that locally, your package.json will make no mention of the version to pick but your package-lock.json will pin to the latest available version that satisfies the different required ranges. 

Because we generate emitter-package.json and its lock file automatically in emitter CI builds, we can't use lock file pinning alone, so we came up with the heuristic of "any dependency mentioned as both a peer dependency and dev dependency must be doing peer dependency pinning, so we'll forward that pin on to emitter-package.json"

This is codified in: [eng/common/scripts/typespec/New-EmitterPackageJson.ps1
](https://github.com/Azure/azure-sdk-for-rust/blob/feature/track2/eng/common/scripts/typespec/New-EmitterPackageJson.ps1#L33)

package.json:
```json
{
  "name": "@azure-tools/typespec-foo",
  "version": "0.35.1",
  "peerDependencies": {
    "@azure-tools/typespec-azure-core": ">=0.46.0 <1.0.0",
    "@azure-tools/typespec-azure-resource-manager": ">=0.46.0 <1.0.0",
    "@typespec/compiler": ">=0.60.0 <1.0.0",
    "another-package": "2.3.4"
  },
  "devDependencies": {
    "@azure-tools/typespec-azure-core": "~0.46.0",
    "@azure-tools/typespec-azure-resource-manager": "~0.46.0",
    "@typespec/compiler": "~0.60.1",
    "something-else": "1.2.3"
  }
}
```

emitter-package.json
```json
{
  "main": "dist/src/index.js",
  "dependencies": {
    "@azure-tools/typespec-foo": "0.35.1"
  },
  "devDependencies": {
    "@azure-tools/typespec-azure-core": "~0.46.0",
    "@azure-tools/typespec-azure-resource-manager": "~0.46.0",
    "@typespec/compiler": "~0.60.1",
  }
}
```

As far as which packages should be included, you can run `npm ls -a`.  If you have any open peer dependencies, it will fail and list them.

Reply:
see: https://github.com/Azure/azure-sdk-for-python/blob/main/eng/emitter-package.json

Reply:
I have a PR out that will make emitter-package.json peer pinning more explicit in the emitter's package.json file:
https://github.com/Azure/azure-sdk-tools/pull/9147

Rather than specifying both peerDependencies and devDependencies in the emitter's package.json you pin the dev dependency and add the package name to a new section:

```json
{
  "name": "@azure-tools/typespec-python",
  "version": "0.35.1",
  "devDependencies": {
    "@azure-tools/typespec-azure-core": "~0.46.0",
    "@azure-tools/typespec-azure-resource-manager": "~0.46.0",
    "@typespec/compiler": "~0.60.1",
    "something-else": "1.2.3"
  },
  "azure-sdk/emitter-package-json-pinning": [
    "@azure-tools/typespec-azure-core",
    "@azure-tools/typespec-azure-resource-manager",
    "@typespec/compiler"
  ]
}
```

This avoids redeclaring transitive peer dependencies, it's less heuristic and it's more declarative.

Reply:
@hallipr so is this PR blocked until the emitter's package.json is updated to this format?

Reply:
@jhendrixMSFT I see several packages that also have ranges in devDependencies: https://github.com/Azure/typespec-rust/blob/a8c4021148d50dc3f54a44f7fd3489568c1e9dba/packages/typespec-rust/package.json#L64-L68 I think we should be pinning to a specific version according to what @hallipr mentioned.

Reply:
I thought the comment was for `emitter-package.json`? Other emitters use the same range which is desirable in order to test emitters against dev versions.

@hallipr can we get clarification on this?

Reply:
@hallipr if this is a new requirement, can we merge as is and then update the emitter-package.json once your new script is updated?

Reply:
`emitter-package.json` should pin all transitive peer dependency ranges as dev dependencies to the same version that was used when testing the emitter.  The only way to reliably get the same versions in package-lock.json and emitter-package-lock.json is to be specific in package.json's devDependencies and automate the creation of emitter-package.json.

If specific version pinning in devDependencies is too much of a headache, we could consider pulling the specific version from `package-lock.json` rather than `package.json`'s devDependencies when creating `emitter-package.json`

Reply:
@jhendrixMSFT thoughts on pinning specific versions in devDeps?

I see c# and python pinning to specific versions already: 
C#: https://github.com/Azure/autorest.csharp/blob/d3615842dd87f222debf00bb1ca3beac0e953942/src/TypeSpec.Extension/Emitter.Csharp/package.json#L54-L81
Python: https://github.com/Azure/autorest.python/blob/333bc8e4b352f487ea24159fde23474fc3679e20/packages/typespec-python/package.json#L67-L92

Reply:
@hallipr have you already updated you script? since last time I ran it, it still just pulled in the peerDeps...otherwise, we'll have to manually update emitter-package.json to unblock this pr

Reply:
I would manually update the emitter-package.json, but use static pinning instead of ranged.

Reply:
Done! Will dismiss the request for changes